### PR TITLE
Add XRaySQ Module

### LINKED ISSUE #42: X-Ray structure factors (new Module)
Add complementary XRaySQ to sit alongside NeutronSQ. Implement general atomic form factors (and/or others) in separate Data structures.

### LINKED ISSUE #322: Add system test for x-ray F(Q) calculation
**Describe the feature / issue**
Add a system test to cover calculation of x-ray weighted structure factors and inverted potential data.

### PR DESCRIPTION
**Pull request by trisyoungs**
This PR adds a new module XRaySQ to calculate x-ray weighted structure factors.

An accompanying system test checks calculation of x-ray structure factors for liquid water.

Closes #42.
Closes #322.

### CODE REVIEW

File: src/modules/xraysq/process.cpp [hunk 1]
```diff
@@ -0,0 +1,479 @@
...
+                            Data1D atomicData = unweightedsq.partial(i, i);
+                            atomicData.values() = weights.formFactor(i, atomicData.constXAxis());
+                            Data1DExportFileFormat exportFormat(
+                                fmt::format("{}-{}-{}.form", uniqueName(), cfg->niceName(), at1.atomTypeName()));
+                            if (!exportFormat.exportData(atomicData))
+                                return procPool.decideFalse();
+                            procPool.decideTrue();
+                        }
+                        else if (!procPool.decision())
+                            return false;
```
Comment (rprospero) [inline src/modules/xraysq/process.cpp:396]: More a question than a comment, but is using `return` to as a substitute for `continue` feel intuitive and ergonomic?  I've had an idea for using a tagged union to allow performing continues or early returns from the process, but I'm not sure if it's worth the hassle.

**Review by rprospero — APPROVED** [review]
Just a couple of minor notes, but otherwise everything looks good.

File: src/classes/xrayweights.h [hunk 2]
```diff
@@ -106,6 +106,10 @@ class XRayWeights : public GenericItemBase
     double weight(int typeIndexI, int typetypeIndexJ, double Q) const;
     // Return full weighting for types i and j (ci * cj * f(i,Q) * F(j,Q) * [2-dij]) over supplied Q values
     Array<double> weight(int typeIndexI, int typeIndexJ, const Array<double> &Q) const;
+    // Calculate and return Q-dependent average squared scattering (<b>**2) for supplied Q values
+    Array<double> boundCoherentSquareOfAverage(const Array<double> &Q) const;
```
[the remark this answers is not in the document]
Comment (trisyoungs) [reply src/classes/xrayweights.h:110]: Agreed. Touches a lot of places, but should be straightforward enough. And certainly worth a performance check before and after!  We can discuss in sprint planning.

File: src/modules/xraysq/process.cpp (hunk 1 above)
Comment (trisyoungs) [reply src/modules/xraysq/process.cpp:393]: @rprospero Now corrected to use `for_each_pair_early`. As regards to the name of this function going forward, perhaps `try_each_pair` fits the bill?

**Pull request merged**